controllers/control/cassandratask_controller.go

```diff
@@ -0,0 +1,480 @@
+}
+
+/**
+	Jobs executing something on every pod in the datacenter
+**/
+const (
+	podJobIdAnnotation     = "control.k8ssandra.io/job-id"
+	podJobStatusAnnotation = "control.k8ssandra.io/job-status"
+	podJobHandler          = "control.k8ssandra.io/job-runner"
+	jobHandlerMgmtApi      = "management-api"
+
+	podJobCompleted = "COMPLETED"
+	podJobError     = "ERROR"
+	podJobWaiting   = "WAITING"
+)
+
+var (
+	// TODO This should be per Datacenter
+	jobRunner         chan int = make(chan int, 1)
+	jobRunningRequeue          = time.Duration(10 * time.Second)
+)
+
+func callCleanup(nodeMgmtClient httphelper.NodeMgmtClient, pod *corev1.Pod, args map[string]string) (string, error) {
+	return nodeMgmtClient.CallKeyspaceCleanup(pod, -1, "", nil)
+}
+
+func callCleanupSync(nodeMgmtClient httphelper.NodeMgmtClient, pod *corev1.Pod, args map[string]string) error {
+	return nodeMgmtClient.CallKeyspaceCleanupEndpoint(pod, -1, "", nil)
+}
+
+func cleanup() TaskConfiguration {
+	return TaskConfiguration{
+		AsyncFeature: httphelper.AsyncSSTableTasks,
+		AsyncFunc:    callCleanup,
+		SyncFunc:     callCleanupSync,
+	}
+}
+
+func callRebuild(nodeMgmtClient httphelper.NodeMgmtClient, pod *corev1.Pod, args map[string]string) (string, error) {
+	return nodeMgmtClient.CallDatacenterRebuild(pod, args["source_datacenter"])
+}
+
+func rebuild(args map[string]string) TaskConfiguration {
+	return TaskConfiguration{
+		AsyncFeature: httphelper.Rebuild,
+		AsyncFunc:    callRebuild,
+		Arguments:    args,
+	}
+}
+
+func (r *CassandraTaskReconciler) getDatacenterPods(ctx context.Context, dc *cassapi.CassandraDatacenter) ([]corev1.Pod, error) {
+	var pods corev1.PodList
+
+	if err := r.Client.List(ctx, &pods, client.InNamespace(dc.Namespace), client.MatchingLabels(dc.GetDatacenterLabels())); err != nil {
+		return nil, err
+	}
+
+	return pods.Items, nil
+}
+
+func (r *CassandraTaskReconciler) reconcileEveryPodTask(ctx context.Context, dc *cassapi.CassandraDatacenter, taskConfig TaskConfiguration) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	// TODO Should we track the pods with UID of the task? So we know this pod has participated..
+
+	// We sort to ensure we process the dcPods in the same order
+	dcPods, err := r.getDatacenterPods(ctx, dc)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	sort.Slice(dcPods, func(i, j int) bool {
+		rackI := dcPods[i].Labels[cassapi.RackLabel]
+		rackJ := dcPods[j].Labels[cassapi.RackLabel]
+
+		if rackI != rackJ {
+			return rackI < rackJ
+		}
+
+		return dcPods[i].Name < dcPods[j].Name
+	})
+
+	nodeMgmtClient, err := httphelper.NewMgmtClient(ctx, r.Client, dc)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for idx, pod := range dcPods {
+		features, err := nodeMgmtClient.FeatureSet(&pod)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if pod.Annotations == nil {
+			pod.Annotations = make(map[string]string)
+		}
+
+		if podJobId, found := pod.Annotations[podJobIdAnnotation]; found {
```

Comment:
I have been testing your branch with my k8ssandra-operator [branch](https://github.com/jsanda/k8ssandra-operator/tree/add-dc-dev) and hit some issues. I am testing adding a new DC to an existing cluster. I first tested a with a single-node DC. I grepped Cassandra's logs to confirm that the rebuild operation. Yay for that!

I then decided to try my test with 2 nodes. When the new DC, dc2, comes up in this case, cleanup runs first. When cleanup finishes, the job id and job status annotations are still present on the pods. The rebuild task is created and reported as finished but not rebuild operations are actually performed. This is because at line 354 we check for the job id annotation. It is already present from cleanup, so we assume that the job is done. 

Once I realized what was going on, I thought I might have a quick fix by simply adding a statement to delete the job id annotation at line 384. Unfortunately it leads to additional problems.

This situation made it abundantly clear that there needs to be an integration for this, i.e., two tasks run in succession.

Removing both the job id and job status annotations when the job completes won't work either because on a subsequent reconciliation we won't be able to determine if the job has already run on that pod. I think that storing the completed pods in the CassandraTask status would make this easier.

Reply:
I think I'll handle this the same way as Kubernetes handles affinity status in annotations (json-serialized job information).

Reply:
If we update TaskConfig to include the task name, then we might be able to add a check to see if both the job id and job status are set but belong to a different task other than the current one, then we would know we can proceed with execution. I can try to test this out today.

Reply:
Well, I did get it to run twice in the test at least with my approach. The current commit is a bit hacky, but I don't have the energy to make it pretty tonight - I'll continue tomorrow. But it seems to pass the test with two tasks in the same datacenter at least.

Reply:
Thanks for pushing the commit. I tested and it didn't resolve the issue.

Reply:
Hmm, I can't see them clashing at least on the annotations part:

```console
Annotations:  control.k8ssandra.io/job-4d7dc3a5-68c4-40cc-9136-cd9b3daac215:
                {"jobId":"448f3370-2aab-42de-8262-257911379814","jobStatus":"COMPLETED","jobHandler":"management-api"}
              control.k8ssandra.io/job-93af73b5-12aa-446e-9e12-720a85227e66:
                {"jobId":"964845be-a7dd-4c1c-977c-08988d8dbf1c","jobStatus":"COMPLETED","jobHandler":"management-api"}
```

Reply:
I also see that the cleanup endpoint is called multiple times. Hmm..